src/inmanta/data/__init__.py

```diff
@@ -5507,123 +5515,254 @@ async def get_resources_for_version(
+            {projection_selectors}
+        FROM reference_model
+        LEFT JOIN model_pairs ON true
+        LEFT JOIN LATERAL (
+            -- DISTINCT because join results in two null rows for shared set. ORDER BY below ensures we keep the latest one
+            SELECT DISTINCT ON (rs_new.name, rs_old.name)
+                -- if the set exists in only one of the models, report that one
+                COALESCE(rs_new.name, rs_old.name) AS name,
+                -- we're only interested in the latest id, or the absence of it
+                rs_old.resource_set_id AS old_resource_set_id,
+                rs_new.resource_set_id
+            FROM (SELECT * FROM rs_with_name WHERE model = model_pairs.old) AS rs_old
+            FULL JOIN (SELECT * FROM rs_with_name WHERE model = model_pairs.new) AS rs_new
+                ON rs_old.name = rs_new.name
+            -- filter out all sets that remained unchanged
+            WHERE rs_new.resource_set_id IS DISTINCT FROM rs_old.resource_set_id
+            -- make sure SELECT DISTINCT matches the new id for the shared set, if it exists
+            ORDER BY rs_new.name, rs_old.name, rs_new.resource_set_id NULLS LAST
+        ) AS diff
+        ON true
+        LEFT JOIN {cls.table_name()} as r
+            ON r.environment = $1
+            AND r.resource_set_id = diff.resource_set_id
+        ORDER BY model_pairs.new, diff.name, r.resource_id
         """
-        resource_records = await cls._fetch_query(query, environment, version, connection=connection)
-        resources = [dict(record) for record in resource_records]
-        return resources
-
-    @classmethod
-    async def get_latest_version(cls, environment: uuid.UUID, resource_id: ResourceIdStr) -> Optional["Resource"]:
-        resources = await cls.get_list(
-            order_by_column="model", order="DESC", limit=1, environment=environment, resource_id=resource_id
+        resource_records = await cls._fetch_query(
+            query,
+            cls._get_value(environment),
+            cls._get_value(since),
+            connection=connection,
         )
-        if len(resources) > 0:
-            return resources[0]
-        return None
+
+        assert resource_records
+        if not resource_records[0]["exists"]:
+            raise PartialBaseMissing()
+        if len(resource_records) == 1 and resource_records[0]["version"] is None:
+            # LEFT JOIN with model_pairs resulted in None row => no new versions
+            return []
+
+        type Model = tuple[int, inmanta.types.ResourceSets]
+        result: list[Model] = []
+        version: int
+        records: Iterator[asyncpg.Record]
+        for version, records in itertools.groupby(resource_records, key=lambda r: r["version"]):
+            spy: Sequence[asyncpg.Record]
+            spy, records = more_itertools.spy(records)
+            assert spy  # groupby can not produce empty groups
+            if spy[0]["empty_model"]:
```

Comment:
is `spy` here mostly to make mypy happy? I would imagine `records[0]` is the same thing, right?

Reply:
No, `records` is an `Iterator`, so there is no such thing as `records[0]`, and `next(records)` would consume the value.